### PR TITLE
Fuse partial sum accumulation into TGMM kernel

### PR DESCRIPTION
**Pull request by Shuwen-Fang**
# Description
This is a follow up to [PR](https://github.com/AI-Hypercomputer/maxtext/pull/4221), see previous PR for more context on motivation of this change. 

Changes in this PR                                                                                               
  1. Backward TGMM Kernel
      • Added optional  partial_sum  input tensor and block specs to  tgmm_v2() .      
      • Updated VMEM budget calculations ( calculate_tgmm_tiling ) to allocate double- 
      buffered space for incoming partial sum tiles.                                   
      • Updated  tgmm_inner_kernel()  to accumulate incoming partial sum tiles in MXU  
      registers ( acc += tiled_ps_ref ) on expert group transitions.
      • Added zero-token expert group handling ( group_size == 0 ) via  jnp.where  to  
      ensure empty experts preserve incoming gradient partial sums rather than being   
      overwritten with zero.
  2. Kernel & Integration Test Suite
      • Unit Tests ( tests/unit/pallas_mosaic_tpu_v2_kernel_test.py ): Migrated Tokamax
      test suite into MaxText and expanded coverage for partial sum feature
   
FIXES: b/522351030
FIXES: b/512156666

# Tests

Ran pytest src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_gmm_kernel_test.py

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tests/unit/pallas_mosaic_tpu_v2_kernel_test.py [hunk 1]
```diff
@@ -257,26 +261,32 @@ def test_gmm_basic(self, batch_size, in_size, out_size, num_groups, has_bias, ha
...
-    key1, key2 = jax.random.split(key, 2)
+    key1, key2, key3 = jax.random.split(key, 3)
     lhs = jax.random.normal(key1, (batch_size, in_size), dtype=jnp.bfloat16)  # [m, k]
     grad = jax.random.normal(key2, (batch_size, out_size), dtype=jnp.bfloat16)  # [m, n]
     group_sizes = get_group_sizes(batch_size, num_groups)
     # if batch_size=128, num_groups=3, an example group_size is
     # group_sizes=Array([14, 14, ..., 7]).
     group_offset = jnp.array(group_offset, dtype=jnp.int32)
 
+    ps = None
```
Comment (gobbleturk) [inline tests/unit/pallas_mosaic_tpu_v2_kernel_test.py:281]: hmm there is something I am not understanding - how was the has_partial_sum=True case passing before this PR?

**Review by gobbleturk — APPROVED** [review]

File: tests/unit/pallas_mosaic_tpu_v2_kernel_test.py (hunk 1 above)
Comment (Shuwen-Fang) [reply tests/unit/pallas_mosaic_tpu_v2_kernel_test.py:281]: It's because we didn't update the reference tgmm, so it was just a no-op parameter

File: src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_tgmm_kernel.py [hunk 2]
```diff
@@ -430,7 +439,7 @@ def out_index_map(self, n_id: jax.Array, k_id: jax.Array, gm_id: jax.Array):
 
 def generate_tgmm_block_specs(
     metadata_ref: gmm_v2.MetadataRef, cfgs: gmm_v2.GmmConfigs
-) -> Tuple[Tuple[pl.BlockSpec, OperandRef], pl.BlockSpec]:
+) -> Tuple[Tuple[pl.BlockSpec, ...], pl.BlockSpec]:
```
Comment (NuojCheng) [inline src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_tgmm_kernel.py]: can we make it `Tuple[Tuple[pl.BlockSpec, OperandRef, pl.BlockSpec | None], pl.BlockSpec]`

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_tgmm_kernel.py (hunk 2 above)
Comment (Shuwen-Fang) [reply src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_tgmm_kernel.py]: updated
